Redis JsonCodec use ObjectMapper without jsr310

### Describe the bug

JsonCodec use ObjectMapper from DatabindCodec.mapper(). Received mapper contains only io.vertx.core.json.jackson.VertxModule.
https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/datasource/codecs/Codecs.java#L64

As a result, if a list of objects is cached that contains a date/time, when trying to deserialize it, an exception is thrown that there is no jsr310 module.

ObjectMapper in Arc contains next modules by default
- VertxTypes
- jackson-datatype-jsr310
- com.fasterxml.jackson.datatype.jdk8.Jdk8Module
- jackson-module-parameter-names

How it can be fixed:
1. Call findAndRegisterModules
`this.mapper = DatabindCodec.mapper().findAndRegisterModules()`
In this case, `findAndRegisterModules` will be called every time. The `VertxTypes` module will be missing.

2. Get ObjectMapper from Arc
`this.mapper = Arc.container().instance(ObjectMapper.class).orElse(DatabindCodec.mapper())`
In this case, the `io.vertx.core.json.jackson.VertxModule` module will be missing.

I think the first option is more correct, since there may be places where the ObjectMapper from `DatabindCodec` is used. And enriching it with the jsr310 module will not be superfluous. And put it in a static field so as not to call `findAndRegisterModules` every time.
But here you know better.


### Expected behavior

Can cached list of objects this date/time

### Actual behavior

Thrown exception
```
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Java 8 date/time type `java.time.LocalDateTime` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 19] (through reference chain: java.util.ArrayList[0]->org.example.Item["localDateTime"])
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67)
	at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1888)
	at com.fasterxml.jackson.databind.deser.impl.UnsupportedTypeDeserializer.deserialize(UnsupportedTypeDeserializer.java:48)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:310)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:361)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:246)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:30)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:342)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4917)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3938)
	at io.quarkus.redis.datasource.codecs.Codecs$JsonCodec.decode(Codecs.java:83)
	... 25 more
```

### How to Reproduce?

[redis-demo.zip](https://github.com/user-attachments/files/18021561/redis-demo.zip)


### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

3.17.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/redis-client/deployment/src/test/java/io/quarkus/redis/deployment/client/datasource/QuarkusObjectMapperTest.java
+++ b/extensions/redis-client/deployment/src/test/java/io/quarkus/redis/deployment/client/datasource/QuarkusObjectMapperTest.java
@@ -1,0 +1,134 @@
+package io.quarkus.redis.deployment.client.datasource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.UUID;
+
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.fasterxml.jackson.databind.ser.std.StdSerializer;
+
+import io.quarkus.jackson.ObjectMapperCustomizer;
+import io.quarkus.redis.datasource.RedisDataSource;
+import io.quarkus.redis.datasource.hash.HashCommands;
+import io.quarkus.redis.deployment.client.RedisTestResource;
+import io.quarkus.test.QuarkusUnitTest;
+import io.quarkus.test.common.QuarkusTestResource;
+
+@QuarkusTestResource(RedisTestResource.class)
+public class QuarkusObjectMapperTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
+            .setArchiveProducer(
+                    () -> ShrinkWrap.create(JavaArchive.class).addClass(CustomCodecTest.Jedi.class).addClass(
+                            CustomCodecTest.Sith.class)
+                            .addClass(CustomCodecTest.CustomJediCodec.class).addClass(CustomCodecTest.CustomSithCodec.class))
+            .overrideConfigKey("quarkus.redis.hosts", "${quarkus.redis.tr}");
+
+    @Inject
+    RedisDataSource ds;
+
+    @Test
+    public void test() {
+        String key = UUID.randomUUID().toString();
+        HashCommands<String, String, List<Person>> h = ds.hash(new TypeReference<>() {
+
+        });
+        h.hset(key, "test", List.of(new Person("foo", 100)));
+        String stringRetrieved = ds.hash(String.class).hget(key, "test");
+        assertThat(stringRetrieved).isEqualTo("[{\"nAmE\":\"foo\",\"aGe\":100}]");
+        List<Person> peopleRetrieved = h.hget(key, "test");
+        assertThat(peopleRetrieved).singleElement().satisfies(p -> {
+            assertThat(p.getName()).isEqualTo("foo");
+            assertThat(p.getAge()).isEqualTo(100);
+        });
+    }
+
+    // without a custom module, this could not be deserialized as there are 2 constructors
+    public static class Person {
+        private final String name;
+        private final int age;
+
+        public Person(String name, int age) {
+            this.name = name;
+            this.age = age;
+        }
+
+        @SuppressWarnings("unused")
+        public Person(String name) {
+            this.name = name;
+            this.age = 0;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public int getAge() {
+            return age;
+        }
+    }
+
+    @Singleton
+    public static class PersonCustomizer implements ObjectMapperCustomizer {
+
+        @Override
+        public void customize(ObjectMapper objectMapper) {
+            SimpleModule module = new SimpleModule();
+            module.addDeserializer(Person.class, new PersonDeserializer());
+            module.addSerializer(Person.class, new PersonSerializer());
+            objectMapper.registerModule(module);
+        }
+    }
+
+    public static class PersonSerializer extends StdSerializer<Person> {
+
+        protected PersonSerializer() {
+            super(Person.class);
+        }
+
+        @Override
+        public void serialize(Person person, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
+                throws IOException {
+            jsonGenerator.writeStartObject();
+            jsonGenerator.writeStringField("nAmE", person.getName());
+            jsonGenerator.writeNumberField("aGe", person.getAge());
+            jsonGenerator.writeEndObject();
+        }
+    }
+
+    public static class PersonDeserializer extends StdDeserializer<Person> {
+
+        protected PersonDeserializer() {
+            super(Person.class);
+        }
+
+        @Override
+        public Person deserialize(JsonParser jsonParser, DeserializationContext deserializationContext)
+                throws IOException {
+            JsonNode node = jsonParser.getCodec().readTree(jsonParser);
+            String name = node.get("nAmE").asText();
+            int age = (Integer) node.get("aGe").numberValue();
+
+            return new Person(name, age);
+        }
+    }
+}

--- a/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/datasource/codecs/Codecs.java
+++ b/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/datasource/codecs/Codecs.java
@@ -11,9 +11,9 @@ import java.util.stream.Stream;
 import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
+import io.quarkus.vertx.runtime.jackson.QuarkusJacksonJsonCodec;
 import io.vertx.core.buffer.Buffer;
 import io.vertx.core.json.Json;
-import io.vertx.core.json.jackson.DatabindCodec;
 
 public class Codecs {
 
@@ -61,7 +61,7 @@ public class Codecs {
                 };
                 this.clazz = null;
             }
-            this.mapper = DatabindCodec.mapper();
+            this.mapper = QuarkusJacksonJsonCodec.mapper();
         }
 
         @Override

--- a/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/jackson/QuarkusJacksonJsonCodec.java
+++ b/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/jackson/QuarkusJacksonJsonCodec.java
@@ -28,7 +28,7 @@ import io.vertx.core.spi.json.JsonCodec;
  * The difference is that this class obtains the ObjectMapper from Arc in order to inherit the
  * user-customized ObjectMapper.
  */
-class QuarkusJacksonJsonCodec implements JsonCodec {
+public class QuarkusJacksonJsonCodec implements JsonCodec {
 
     private static volatile ObjectMapper mapper;
     // we don't want to create this unless it's absolutely necessary (and it rarely is)
@@ -43,7 +43,7 @@ class QuarkusJacksonJsonCodec implements JsonCodec {
         prettyMapper = null;
     }
 
-    private static ObjectMapper mapper() {
+    public static ObjectMapper mapper() {
         if (mapper == null) {
             synchronized (QuarkusJacksonJsonCodec.class) {
                 if (mapper == null) {

